### ISSUE TITLE
FR: MIDI Edit: toggle SNAP-to-Note ON/OFF

### ISSUE THREAD

**Issue by MoebiusSt** [ThingstoneGmbH/Track8#108]
In MIDI Edit it's a major pain to align the fixed crosshair with a midi note on the "sheet" while using two encoders for X-/Y-navigation. Touch gestures as solution are postponed. The idea is to make navigation from MIDI note to MIDI note "snappy":

In MIDI Edit View the Key TRACK7 is still free for a local toggle setting; and sits perfectly next to a similar toggle.

**SNAP**

if ON any X- or Y-Encoder scroll will put the nearest MIDI Note in that scroll direction under the crosshair. This way we can navigate through all the midi notes a little faster.

Append with limitations, issues, edge cases.

_[#101, idea origin](https://github.com/ThingstoneGmbH/Track8/discussions/101) can be deleted. Discussion continues here, below.

**Comment by ColRad** [comment]
@MoebiusSt 
Implemented in [v1.1.2-beta6](https://github.com/ThingstoneGmbH/Track8/releases/tag/v1.1.2-beta6)

Would be great if you could test this and post any comments/improvements you have.

**Comment by MoebiusSt** [comment]
Current implementation of SNAP in .v1.1.2-beta6 is is not fully satisfactory. 

You jump to the next X-Position, disregarding any Y-snapping. That was not quite what i meant. But it could become a viable solution: if the Y-Encoder WOULD also snap up down in the next step. So, yours would be a two-step approach: snap-x to next or previous MIDI-note-Begin, then Snap-Y up or down with the Y-Encoder if so wanted. This puts focus on fast time-traversal, and against voicing traversal. That could be ok. 

But a two step-approach was not what i had in mind. So i played around with snap-algorithms... 
The TWO-step approach is there too, it's "X-then-Y" mode.

Try them out, please:
**[Track8 SNAP Algo Tester: ...  https://moebiusst.github.io/track8_midi-note-snap/ ](https://moebiusst.github.io/track8_midi-note-snap/)**
**USE ARROW KEYS**

They all have compromises  :( But some like "Directional Bias" feel almost "very good".

You'll find PSEUDO-Code descriptions for each attempt in the toggle box below the view.
All algos that use pixels for distance measuring you would be able adapt to your quantized grid and midi-note-pitch-data via Conversion factors on the y-axis, you wouldn't have to uses pixels.

The "PITCH BIAS"-Button finetunes some algos snap decisions weighting. Lower values 1.5 to 2 priorizes pitch jumps. higher values like 3 or 5 priotizes time jumps.

**Comment by ColRad** [comment]
Wow, awesome preview tool!

So the idea in my implementation was: 
- All notes in the viewport span a custom SNAP grid in X/Y
- X/Y encoders snap to that grid

Even though this needs both encoders to get to a note, imho it feels quite intentional and deterministic.

 What I find confusing in all your algorithms is: 
- Sometimes when jumping forward/backward I don't land on the same note
- When I have two notes (same pitch) following each other, X Encoder movements don't take me to the next note

I think I would prefer following behavior
- X always takes me to the next/previous note with the same pitch (if there are any in the viewport, else to visually closest next/previous note)
- Reversing the direction always lands me on the previous selected note
- There should be a really strong bias for the axis.

One thing to consider: 
We don't need to jump to the note start, a note is also considered selected if the cursor is somewhere inside of a note. 
This could help to avoid large X jumps when turning the Y encoder.

**Comment by MoebiusSt** [comment]
I will further explore. And also answer to your last post. The problem is not solved yet. I think it always needs to jump to Note start, in order to make absolutly clear what happened/which note was meant. Unless we think about "mouse/cursor-over" HIGHLIGHTING  in conjunction with your proposal to NOT jump to Note start) but this would change the entire approach. Right now its: One note, one single point. BUT if it always must jump to Note start (when no highlighting) there is a chance that smarter algorithms could make use of the NOTE-length for better judgement at least – and also to more closely match human intuitive expectancy of the traversal. This might also improve "reversibillity". In other words, right now all algos consider this as a grid of points, but with overlapping notes its more complex than that, and the algos don't account for it yet.

ALSO note: some algos work better in dense notes and some work better in sparse notes .... Especially your two-step apprach will fail in dense notes. The test tool will have to be upgraded with a density slider in order to test that. 

More in a few days or next week.

**Comment by ColRad** [comment]
So even though there is no highlight yet, internally a note is already considered selected when the crosshair is on the note bounds (when you press start/length/velocity/note encoder the note gets rendered orange to show its modified)

Currently you can only delete/replace a note if you toggle record (then all notes get rendered red). Maybe we can change that around a bit: 
- don't allow toggling record in MidiEdit>Note
- A selected note gets rendered orange to make clear it's selected
- The modify state changes from orange -> red
- Replace/Delete is always available when a note is selected (so no need for record)
- Add is always available when no note is selected (so no need for record) 

I'd rather have these changes to highlight that a note is selected than big horizontal jumps to the start of a note. 

Even though Midi notes technically consist of two independent events (press/release). There should not be any overlap of notes (note1press, note2press, note1release, note2release). And Track8 also makes sure that this does not happen (if you insert a short note inside of a long note, the long note gets cut and it's release set one tick before the second notes start)

Just saying, if adding highlighted notes reduces complexity, and makes the snap algorithms feel more intentional, we can absolutely do it.

**Comment by MoebiusSt** [comment]
Cool. Highlighting the note will no longer make necessary to jump to note-start. This way algos can work different and attempt to jump less. We'll see. (With "overlapping" we had a little misunderstanding. With overlapping i meant different notes playing at the same time - like in chords. For example a lower pitched note can start far far to the left of another note from where the cursor currently is, and be held– its bar extending beyond our current note. Most test-algos now would never jump to it, even though from the human perspective clearly the other note lies just a few halfsteps below, the bar obviously in reach, "why doesn't it jump to it, i just went DOWN!!?". The new appraoch would take note lengths into account and just jump "into" it on the y-axis, then highlight it, keeping time, not moving on the x-axis, And for the X-Encoder you wrote "X always takes me to the next/previous note with the same pitch (if there are any in the viewport, else to visually closest next/previous note)". I think i understood you. (its this very strong bias for direction you wanted). Will have to read your last two posts again carefully and think it all through. I will try it out.

**Comment by ColRad** [comment]
All very complicate if one thinks about the details xD

So lets say we have following notes with X being the cursor position

```
D ---                       ----------
C#-----                   ----------- 
C ----------------------   -X--------
```

If we move X-encoder to the left, I would assume to land on the first C (preferably at the start)
If I move Y-Encoder up, I would assume to not change X position, but land inside of the second C#. 

If we have this: 

```
D ---                 ----------
C#-----                   
C ----------------------   -X--------
```
My gut feeling would assume that encoder-Y up would take be to  the middle of D. If this took me to the start of D, a reverse direction would take me to the first C Note, not the second where I started.


Now what about this:
```
D ---                   
C#-----                   
C ----------------------   -X--------
```
Is it justified there to make a big X axis jump to C# on encoder-Y up? If we do and reverse the direction we end up on the first C note. What would be the alternative? Do nothing on encoder-Y up? Fallback to grid?

I think I'd rather have something that is deterministic and reversible in it's directional jumps
than something that always lands be on a note.


I honestly see two algorithms that would do that: 

1. x/y Cross detection from cursor
- On X turns we check for the next/previous note of the same pitch to jump to, if there is no note we jump in grid.
- On Y turns we check for the higher/lower note that extends through the current playhead position to jump to, else we jump  one semitone.

2. Time based X Axis snap
- On X turns we jump to the next/previous note start, if more than one note starts at the same time we choose the one that is in pitch closest to the current curser-Y (Basically the X movement of your Axis priority, but ignoring the viewport) If there is no previous/next note we do nothing. This is basically reminiscent of a tracker approach. 
- Y Turns are only used to select different notes of chords that start at exactly the same time. Viewport is also ignored, all notes are taken into account. If there is no higher/lower note, we do nothing. 


**Edit**
Or we make 2. even simpler: 
We order chord notes from high to low pitch. So the user only has to use the X-Encoder to reach all notes.
So the truly feels like editing midi notes on a Tracker.
Y-Encoder is not snapped and just moves the cursor freely in semitones. 

I think I'd prefer 2. 
This would allow me to use only one encoder (or the arrow keys) to reach all notes. And have my left hand free for the modification encoders.

**Comment by MoebiusSt** [comment]
Collin, please try out the implementation of our latest approach: 
### ‘Time / Harmony’ (default) 

https://moebiusst.github.io/track8_midi-note-snap/

- allows for minimal semitone steps. You can set it from 0 to 12. 
- **Highlighting** is implemented.
- Added a slider for test-note density – experiment with sparse or denser note arrangements.

It is quite good! But note the small highest notes... they are unreachable with low Pitch Gate setting :(

AI Summary:


> **The strict separation of axes solved almost all problems.** The previous 6 algorithms all attempt to find the “best” note in a 2D scoring system — and fail because it is mathematically impossible to find a single scoring function that feels humanly intuitive across all note configurations. Directional Bias came closest, but even there were the diagonal jumps.
> 
> The new approach circumvents this completely:
> - **Left/Right** is a purely 1D problem: next starting point in time, pitch only as a tiebreaker for harmonic intervals. Trivially reversible.
> - **Up/Down** is also a purely 1D problem: next pitch among the currently sounding notes. Trivially reversible. And thanks to the note extension (“sounding at cursor.x”), long-held notes finally become reachable.
> 
> The new algorithm combines the strengths of **‘Axis Priority’** (clear axis separation) with the concept of **note extension**, refined with a slight fine-tuning from **‘Directional Bias’** (pitch proximity as a tiebreaker). 
> 
> The **Reversible Modifier** (a special safeguard that temporarily stores the last source note) is no longer needed at all — reversibility is a natural consequence of the axis separation.
> 
> Another idea I wanted to pursuit, to reintroduce a **view radius** / light cone is also no longer needed, because Up/Down only finds notes that are actually sounding at the current moment — this is a natural, musically meaningful filter.

**Comment by MoebiusSt** [comment]
**About those unreachable Notes:** You can’t hit the short high notes using the Up key, because your X key is never in a position where they sound. And you can’t hit them using the Right key, because the pitch interval is too wide, too far apart, with Pitch Gate = 2.

**There are three possible solutions:**

- Increase the pitch gate — at 12 (default), they can be reached using the right arrow key, but not at 2. This is the simplest fix, but this breaks the strict horizontal navigation.
- Extend Up/Down: include not only notes “sounding at cursor.x”, but also notes whose start is close to cursor.x. E.g. a small time tolerance range: startTick - TOLERANCE <= cursor.x < startTick + duration + TOLERANCE. This would be a second tunable parameter.
- Hybrid approach: If Up/Down cannot find a sounding note, implement a fallback that searches for the note closest in pitch within a certain time radius — similar to the fallback for the pitch gate.

I added the option 3 "Hybrid approach" for those upper(lower short note edge cases.
You will find the mode **"Time / Harmony w. Fallback"** in the Dropdown. All notes are now reachable. With lower PITCH GATE allowance of only 2 and in sparser note arrangements it still often causes irritation, 
But, what do i mope about. With PITCH GATE at middleground 5 it all works pretty fine!
Do you take it from here?

**Comment by ColRad** [comment]
@MoebiusSt 
I released a [v1.1.2-SNAPSHOT](https://github.com/ThingstoneGmbH/Track8/releases/tag/v1.1.2-SNAPSHOT) version with both your Harmony algorithm and my Tracker algorithm for testing. 

The `SNAP` button now cycles through `OFF`-> `HARM SNAP` -> `TRKR SNAP`

**Comment by MoebiusSt** [comment]
I think this is solved. Works intuitive enough with very few mental snags. Offering the second SNAP mode (great idea! works great too) for those who prefer it. Lovely, all.

I see you decided against highlighting? I agree that it seems superflous. But it IS a visual aid. Have you seen two notes that sit next to each other with no gap (horizontally. Also vertically (i know you went the extra mile with that 1px margin, but still)) Highlighting helps to discern, and also to understand what's going on. Doing the SNAP-Y into the inside of a note is seldom, yes, but highlighting would leave NO questions.
 (I see "auto-select" doesn't work out because of the Select-and-Move action.). But highlighting could be a different color.)
But yeah, maybe its not worth the hassle.

Minor nitpicks: 
- the white text (velocity) on light green bar is not readable. (default & dark theme), i suggest a darker green. The current bright green is learned as SELECT in the audio tracks views, and could work as the "highlight" color here (over darker green notes)
- the cursor could use 4px offset for optical alignment
- the cross inside the crosshair... is "blurry" ? Did you try a dot?
- the piano roll bars in the background (black and white key bars) are DARKER in the DEFAULT-Mode than in the DARK-Mode ;)

![Image](https://github.com/user-attachments/assets/1587c4e1-41d9-4890-aff0-950c955b8bb6)

**Comment by ColRad** [comment]
I'll have a look at your nitpicks ;) 

@MoebiusSt would you mind checking https://github.com/ThingstoneGmbH/Track8/issues/106 ? 
What is your opinion on the issues I raised?

**Comment by ColRad** [comment]
@MoebiusSt 
I implemented some of your recommendations in the new [SNAPSHOT](https://github.com/ThingstoneGmbH/Track8/releases/tag/v1.1.2-SNAPSHOT) upload:

  - NOTE mode: changed note default color to the muted color, only the selected note is rendered in the bright, selected color
  - Fixed Cursor offset and replaced center cross with center dot

**Comment by MoebiusSt** [comment]
Awesome!

- In MIDI Overview when i switch tracks with Track-Keys 1-8, the right side OUTPUT indicators and output METERS below flash sometime for a few milliseconds? Is that a bug or a feature (midi necessities?) I have no Midi currently connected.
- in MIDI Editor i expected to click on a MARKER in order to lock it or name it. (same in AUDIO EDIT) This only works in Audio or Midi Overview. (low priority, just to round stuff up)

**Comment by ColRad** [comment]
- I do send Midi Stop note (from ringing out) Events for several exit scenarios. Because I had the issue with some synthesizers that the notes were stuck. Maybe thats too aggressive to also do it on Track selection. 
- Ahh yeah, I didnt want to add this to the Edit screens, because we have some other touch features there (automation/cc/pitchbend). Idk maybe my worries are exaggerated.

**Comment by MoebiusSt** [comment]
in MIDI EDIT 
  if HRM SNAP or TRKR SNAP is on
    while is playing 
     disable SNAP and fall back to normal scroll (because snapping won't let you scoll back during playback)

**Comment by ColRad** [comment]
Added in the [SNAPSHOT from today](https://github.com/ThingstoneGmbH/Track8/releases/tag/v1.1.2-SNAPSHOT)

**Comment by ColRad** [comment]
Implemented in [v1.1.2](https://github.com/ThingstoneGmbH/Track8/releases/tag/v1.1.2)
